### DISCUSSION TITLE
Plot space that crosses 180 degrees

### DISCUSSION THREAD

**Discussion by rebeccaherman1** [SciTools/iris#5450] [Q&A] [answered]
Below is my attempt to plot SST in the Pacific Ocean between 5S and 5N, and between 160W and 80E. 
![Screenshot from 2023-07-25 16-53-50](https://github.com/SciTools/iris/assets/12139022/e59476ce-0eaf-4808-86e0-0ca00f05fe0e)
As you can see, it is plotted from 180E to 180W instead of breaking the plot in the region where there is no data. This causes additional problems because if I try to set the x limits from 160 to -80 in any fashion (i.e. adding 360 to the second number or subtracting from the first), the plot focuses on the region with no data.

If someone could quickly explain to me how to change the longitude breakingpoint of the plot, I'd be much appreciative. I cannot come up with useful search words.

https://github.com/SciTools/iris/blob/2a9e131875be42c39ab4b8fc9956e4a4d59f83a2/lib/iris/quickplot.py#L4

**Comment by HGWright** [comment]
@SciTools/peloton There are a number of potential causes for the graph not working as expected, to pinpoint the problem we would need to recreate the cube or have access to the data. Could you upload a single timestep of the cube?

Thanks for raising this!

**Comment by rebeccaherman1** [comment]
[pacific_example.zip](https://github.com/SciTools/iris/files/12172035/pacific_example.zip)

Thank you!

**Comment by acchamber** [comment]
Can replicate the issue in the latest iris version.

 ```
import iris
import iris.quickplot as qplt

cube = iris.load_cube("pacific_example.nc") 
qplt.pcolormesh(cube)
````
![image](https://github.com/SciTools/iris/assets/68277260/16770fdc-80b6-43b8-bfd2-ce4c5181264f)

If I plot the data manually using a PlateCarree projection I get a contiguous plot as expected 

```
import matplotlib.pyplot as plt
import cartopy.crs as ccrs

ax = plt.axes(projection=ccrs.PlateCarree())
plt.pcolormesh(cube.data,transform=ccrs.PlateCarree())
```

![image](https://github.com/SciTools/iris/assets/68277260/71b7d938-b7af-4bcd-9442-17ca083c7e70)


And the cube doesn't have any coord_system attached so it should be using PlateCarree

```
print(cube.coord_system())
None
print(cube.coord("latitude").coord_system
None
print(cube.coord("longitude").coord_system
None
```

That's my initial ideas, any other thoughts on the cause of this issue?

**Comment by rebeccaherman1** [comment]
Update:
Though I have not changed my code at all, the effect of setting the limits of the plot has changed: 
Now if I set the limits to `[160-360, -80]` it will show me [-180,-80];
and if I set the limits to `[160, 360-80]`, it will show me [160,180],
but the only way to see both at the same time is to not limit the x-axis, in which case the figure still spans from -180 to 180 with a large section of no data in the middle, as before.
(setting the limits to `[160, -80]` still shows just the area with no data)

Any further thoughts in the meantime on how to deal with this issue @HGWright?

**Comment by rcomer** [comment]
I think you need to set up your GeoAxes with a chosen central longitude (default is 0).

```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import iris
import iris.quickplot as qplt

cube = iris.load_cube("pacific_example.nc")
ax = plt.subplot(111, projection=ccrs.PlateCarree(central_longitude=210))
qplt.pcolormesh(cube, axes=ax)
plt.show()
```

![image](https://github.com/SciTools/iris/assets/10599679/95064cc9-a961-4f47-84f3-3273a82413dd)

Arguable Iris could/should be more clever about choosing that for you automatically.

**Comment by bjlittle** [comment]
@SciTools/peloton Do you need further assistance @rebeccaherman1 with this issue?

**Comment by rebeccaherman1** [comment]
Yes, @bjlittle, I think there's more to say.
Thanks @rcomer for the tip. This seems to work well as long as I don't mind having no axes labels on my plot. However, I want the viewer to know what they are seeing. 

I am now avoiding the `subplot` hack and using the following code:
```
    ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=220))
    qplt.pcolormesh(cube, axes=ax) 
    ax = plt.gca()
    ax.coastlines()
```
This produces the following plot:
(I am now plotting a slightly different quantity, but it shouldn't make any practical difference.)

![Screenshot from 2023-08-02 14-48-43](https://github.com/SciTools/iris/assets/12139022/e071db6c-e948-4961-9bee-6c6f03066803)

However I have just realized that this results in relabeling 220E as 0, which means that I can no longer access the original longitude values for adding xticks to the plot, dramatically reducing the advantage of having the data in a cube. I'm working on a workaround for my present code now that I understand the problem, but I do not think this is a general sufficient approach. I agree with @rcomer that Iris should handle this automatically.

**Comment by rcomer** [comment, marked as answer]
Maybe have a look at Cartopy's [gridlines](https://scitools.org.uk/cartopy/docs/latest/reference/generated/cartopy.mpl.geoaxes.GeoAxes.html?highlight=gridlines#cartopy.mpl.geoaxes.GeoAxes.gridlines) method for axis labels.  `xticks` is a Matplotlib thing, and Matplotlib doesn't know anything about Cartopy's special transforms.

```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import numpy as np
import iris
import iris.quickplot as qplt

cube = iris.load_cube("pacific_example.nc")
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=210))
qplt.pcolormesh(cube, axes=ax)

ax.gridlines(draw_labels=True, linewidth=0, xlocs=[160, 180, -160, -140, -120, -100, -80])

plt.show()
```

![image](https://github.com/SciTools/iris/assets/10599679/c65bc7e6-910c-4ba0-a92f-64a6e809a866)

**Comment by rebeccaherman1** [comment]
Thank you, that is helpful. 

I must say that I find the available functions somewhat frustrating. I don't like that `gridlines()` labels 4.5N and S rather than 5N and S, which mark the top and bottom of the plot. If I try to include these ticks forcibly, qplt elongates the plot to include up to 8N and S with no data there. Something similar happens when I try to use `set_extent` instead of the `matplotlib` `xlims`. Any ideas on how to avoid this?

For another plot, I believe I need to use `set_extent` because for some reason the plot fills in all the missing space with the color of the last point, and it also adds white space at the top. 

Without `set_extent`:
![Screenshot from 2023-08-02 17-08-33](https://github.com/SciTools/iris/assets/12139022/1e203fb9-20c5-4bdf-9664-3f3c3c6011d4)

With `set_extent`:
![Screenshot from 2023-08-02 17-08-11](https://github.com/SciTools/iris/assets/12139022/6780517d-a352-4b7d-90e6-c02eb410182d)

**Comment by rcomer** [comment]
I'm not sure about the `set_extent` behaviour, but `set_ylim` seems to work as expected for me.  If you look at the bounds of your latitude coordinate, they only extend to +/- 4.71 degrees, so that will determine the default axis limits.

For your EOF plot, the most likely explanation is that the [circular](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.coords.html#iris.coords.DimCoord.circular) property of your longitude coordinate is set to `True` when it should be `False`.

**Comment by rebeccaherman1** [comment]
Thanks for the explanation and the idea, @rcomer.

> I'm not sure about the set_extent behaviour, but set_ylim seems to work as expected for me. If you look at the bounds of your latitude coordinate, they only extend to +/- 4.71 degrees, so that will determine the default axis limits.

I think I wasn't expecting this because I had called `extract_region` using [-5,5]. Would the bounds be lower because of the grid? Or because they mark the center of the grid point rather than the edge?

> For your EOF plot, the most likely explanation is that the [circular](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.coords.html#iris.coords.DimCoord.circular) property of your longitude coordinate is set to True when it should be False.

This is a property of the cube itself, right? It was a good idea, but I have checked, and the circular property of all components is set to 0. Incidentally, I wondered if this was potentially a reason why I was having trouble plotting the pacific_example, and if setting the coordinates to circular could stand in for setting the center of the plot. However, this does not help. 

However it does have an effect exactly like what I see in the EOF plot, but it doesn't seem to be the cause in the EOF case. Any ideas of other fields that could have the same effect?

**Comment by ESadek-MO** [comment]
@SciTools/peloton We are converting to a discussion as it seems to have moved on from the original question.

**Comment by HGWright** [comment]
@SciTools/peloton There no longer seems to be any activity here, and a lot of the questions have been answered. So happy to close this. Please raise more issues/discussions when you need it. Thanks!
